File: java/leetcode/containerWithMostWater.java
/*
 * Container With Most Water
 *
 * https://leetcode.com/problems/container-with-most-water
 */

import java.lang.Math;

class Solution {
	public int maxArea(int[] height) {
		return this.maxAreaHelper(height, 0, height.length - 1);
	}

	public int maxAreaHelper(int[] height, int i, int j) {
		if (j <= i)
			return 0;

		int area = (j - i) * Math.min(height[i], height[j]);

		if (height[i] < height[j])
			return Math.max(area, this.maxAreaHelper(height, i + 1, j));

		return Math.max(area, this.maxAreaHelper(height, i, j - 1));
	}
}

File: python/leetcode/container_with_most_water.py
"""Container With Most Water

https://leetcode.com/problems/container-with-most-water/
"""

class Solution:
    def maxArea(self, height: List[int]) -> int:
        return self.maxAreaHelper(height, 0, len(height) - 1)

    def maxAreaHelper(self, height: List[int], i: int, j: int) -> int:
        if j <= i:
            return 0

        area = (j - i) * min(height[i], height[j])

        if height[i] < height[j]:
            return max(area, self.maxAreaHelper(height, i + 1, j))

        return max(area, self.maxAreaHelper(height, i, j - 1))
